### PR TITLE
Alerting: useAbility hook for alert rules

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

Continuation of https://github.com/grafana/grafana/pull/72626; this PR introduces a `useAlertRuleAbility` hook to map user actions to both `supported` and `allowed` properties taking both RBAC and alert rule source context in to account.

**Why do we need this feature?**

See PR linked above for rationale, a lot of code in alerting is hard to parse because assessing if a particular action is supported by the context _and_ RBAC checks are decoupled and somewhat all over the place. 

Using a single set of hooks to managed these makes it easier to reason about our code and audit access control.

**Special notes for your reviewer:**

~Currently in draft until I can wrap up some of the TODOs and have added some tests / done some manual testing.~

For anyone reviewing this I'd recommend spending time on checking for regressions too both in code and in manual testing :)

### CODE REVIEW

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 1]
```diff
@@ -49,53 +58,137 @@ export enum AlertmanagerAction {
...
-    [AlertSourceAction.DeleteAlertRule]: [AlwaysSupported, ctx.hasPermission(AccessControlAction.AlertingRuleDelete)],
-    // -- External alert rules (Mimir / Loki / etc) --
-    // for these we only have "read" and "write" permissions
-    [AlertSourceAction.CreateExternalAlertRule]: [
+/**
+ * This one will check for alerting abilities that don't apply to any particular alert source or alert rule
+ */
+export const useAlertingAbilities = (): Abilities<AlertingAction> => {
+  const canExportGMARules =
+    ctx.hasPermission(AccessControlAction.AlertingProvisioningRead) ||
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useAbilities.ts:102]: Why are we changing the permissions in there?. For exporting users only need `AccessControlAction.AlertingRuleRead`.

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 2]
```diff
@@ -49,53 +58,137 @@ export enum AlertmanagerAction {
...
+  Duplicate = 'duplicate-alert-rule',
+  View = 'view-alert-rule',
+  Update = 'update-alert-rule',
+  Delete = 'delete-alert-rule',
+  Explore = 'explore-alert-rule',
+  Silence = 'silence-alert-rule',
+  ModifyExport = 'modify-export-rule',
+}
+
+// this enum lists all of the actions we can perform within alerting in general, not linked to a specific
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useAbilities.ts:72]: nice ! 💯 these actions separated in two different lists.

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 3]
```diff
@@ -49,53 +58,137 @@ export enum AlertmanagerAction {
...
+ */
+export function useAlertRuleAbility(rule: CombinedRule, action: AlertRuleAction): Ability {
+  const abilities = useAllAlertRuleAbilities(rule);
+
+  return useMemo(() => {
+    return abilities[action];
+  }, [abilities, action]);
+}
+
+export function useAlertRuleAbilities(rule: CombinedRule, actions: AlertRuleAction[]): Ability[] {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useAbilities.ts:149]: I don't see any place where we are using this hook. Is like an alternative for the  `useAlertRuleAbility` if you want to use a list of actions at once?

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 4]
```diff
@@ -49,53 +58,137 @@ export enum AlertmanagerAction {
...
+  // if a rule is either provisioned or a federated rule, we don't allow it to be removed or edited
+  const immutableRule = isProvisioned || isFederated;
+
+  // TODO refactor this hook maybe
+  const {
+    isEditable,
+    isRemovable,
+    isRulerAvailable = false,
+    loading,
+  } = useIsRuleEditable(rulesSourceName, rule.rulerRule);
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useAbilities.ts:173]: yes, I think we could extract this permissions logic that are inside this `useIsRuleEditable` and use it here. So, this `useIsRuleEditable` would only return if the rule is editable but don't check any permission.  I think all permissions should be checked in here.

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 5]
```diff
@@ -202,7 +295,30 @@ export function useAlertmanagerAbilities(actions: AlertmanagerAction[]): Ability
...
 
-export function useAlertSourceAbility(action: AlertSourceAction): Ability {
-  const abilities = useAlertSourceAbilities();
-  return useMemo(() => abilities[action], [abilities, action]);
+/**
+ * We don't want to show the silence button if either
+ * 1. the user has no permissions to create silences
+ * 2. the admin has configured to only send instances to external AMs
+ */
+export function useCanSilence(rulesSource: RulesSource): [boolean, boolean] {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useAbilities.ts:303]: what about making this hook private to avoid using it directly ?

File: public/app/features/alerting/unified/hooks/useAbilities.ts [hunk 6]
```diff
@@ -49,53 +58,137 @@ export enum AlertmanagerAction {
...
+    [AlertingAction.DeleteExternalAlertRule]: [
       AlwaysSupported,
       ctx.hasPermission(AccessControlAction.AlertingRuleExternalWrite),
     ],
   };
+};
+
+export const useAlertingAbility = (action: AlertingAction): Ability => {
+  const allAbilities = useAlertingAbilities();
+  return allAbilities[action];
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/hooks/useAbilities.ts:135]: We should – good catch!

File: public/app/features/alerting/unified/hooks/useAbilities.ts (hunk 3 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/hooks/useAbilities.ts:149]: That was the original intention but it seemed to add a larger surface for mocking (we should do less of that, but that's another conversation).

I tried comparing one single `useAlertRuleAbilities()` hook to multiple `useAlertRuleAbility()` hook calls but both read equally good. With the `useAlertRuleAbilities()` you have to make sure to desctruct from the Array in the right index or it will cause bugs.

I'm thinking of just not doing any plural hooks, I've left this one in in case we change our minds but it's probably safer to reduce our code surface and choose the hook with the least potential for bugs.

File: public/app/features/alerting/unified/hooks/useAbilities.ts (hunk 5 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/hooks/useAbilities.ts:303]: Definitely, this hook shouldn't be called directly but use `useAlertRuleAbility(rule, AlertRuleAction.Silence)` instead

File: public/app/features/alerting/unified/hooks/useAbilities.ts (hunk 1 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/hooks/useAbilities.ts]: Ok so it turns out after a conversation that there are slight inconsistencies between provisioning permissions of alert rules and Alertmanager entities. I'm reverting the check back for now to be consistent with the backend permission checks but we should sort those out at some point.

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM! 🚀
nit: `useCanSilence` is still public, I think we should make it private to avoid using it directly 👀
